### PR TITLE
Changes/fixes to serverless pattern

### PR DESCRIPTION
**Pull request by mialarconchong**
*Issue #, if available:* N/A

*Description of changes:* Changes to Lambda function code and minor changes to stack

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cloudtrail-lambda-dynamo-cdk/README.md [hunk 1]
```diff
@@ -29,6 +28,10 @@ s.
     ```
     npm install
     ```
+1. Open the lambda function cloudtrail-lambda-dynamo-cdk/src/lib/lambda/object_tag_checker/index.py, and on line 17 replace the example keys with your required keys:
+    ```
+    required_keys = {"Key1", "Key2", "Key3", "Key4"}
+    ```
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/README.md:34]: Can you please explain what these resource keys are. Are they S3 object keys?

Comment (mialarconchong) [reply cloudtrail-lambda-dynamo-cdk/README.md]: Yes they're S3 object keys, added this in the latest commit as well

File: cloudtrail-lambda-dynamo-cdk/README.md [hunk 2]
```diff
@@ -29,6 +28,10 @@ s.
     ```
     npm install
     ```
+1. Open the lambda function cloudtrail-lambda-dynamo-cdk/src/lib/lambda/object_tag_checker/index.py, and on line 17 replace the example keys with your required S3 object keys:
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/README.md:31]: should be `resource_tag_checker` folder for Lambda. and Line 40

Comment (julianwood) [reply cloudtrail-lambda-dynamo-cdk/README.md:31]: Suggest also wording this as an optional for your real life use-case. For the test, you don't need to modify these. If I understand correctly it's not S3 object keys, its the S3 object key *tags* that are being checked.  So, speficy that you can amend the tags associated with the S3 objects that you want to check

File: cloudtrail-lambda-dynamo-cdk/README.md [hunk 3]
```diff
@@ -40,25 +43,31 @@ s.
 ## Testing
 
 Once the CDK stack has deployed successfully, you can take the following steps to ensure the pattern is working appropriately:
-1. Using the AWS CLI, create a bucket using the following command:
+1. Using the AWS CLI, upload the test_file.txt found in the src folder to the S3 bucket of your choosing using the following command:
     ```
-    aws s3api create-bucket --bucket {bucket-name} --region {your AWS region}
+    aws s3 cp test/test_file.txt s3://<bucket-name>/est/test_file1.txt
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/README.md:48]: Extra `/est` in the file path.

File: cloudtrail-lambda-dynamo-cdk/src/lib/tag-compliance-stack.ts [hunk 4]
```diff
@@ -11,24 +12,21 @@ export class TagComplianceStack extends cdk.Stack {
...
       stream: dynamodb.StreamViewType.NEW_IMAGE,
-      removalPolicy: cdk.RemovalPolicy.DESTROY // can be adjusted as deemed necessary
+      removalPolicy: cdk.RemovalPolicy.DESTROY 
     });
 
     // random generated number appended to bucket name, to create unique name
-    const generatedNum = Math.ceil(Math.random() * 10);
+    const generatedNum = Math.ceil(Math.random() * 100);
 
-    const bucket = new s3.Bucket(this, 'resourceCreationLogs', {
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/src/lib/tag-compliance-stack.ts:30]: Suggest also outputting the bucket name when the stack is deployed so its easier to find out what it is for testing

File: cloudtrail-lambda-dynamo-cdk/src/lib/tag-compliance-stack.ts [hunk 5]
```diff
@@ -11,24 +12,21 @@ export class TagComplianceStack extends cdk.Stack {
...
         type: dynamodb.AttributeType.STRING
       },
       stream: dynamodb.StreamViewType.NEW_IMAGE,
-      removalPolicy: cdk.RemovalPolicy.DESTROY // can be adjusted as deemed necessary
+      removalPolicy: cdk.RemovalPolicy.DESTROY 
     });
 
     // random generated number appended to bucket name, to create unique name
-    const generatedNum = Math.ceil(Math.random() * 10);
+    const generatedNum = Math.ceil(Math.random() * 100);
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/src/lib/tag-compliance-stack.ts:26]: Just a suggestion. Another way to generate uniqueness is to use the account number.

File: cloudtrail-lambda-dynamo-cdk/src/lib/lambda/populate_dynamo/index.py [hunk 6]
```diff
@@ -1,87 +1,59 @@
...
-            object_key = record['s3']['object']['key']
-            
-            response = s3_client.get_object(Bucket=bucket_name, Key=object_key)
-            log_data = gzip.decompress(response['Body'].read()).decode('utf-8')
-            cloudtrail_logs = json.loads(log_data)['Records']
-            
+
+            object_key = record["s3"]["object"]["key"]
+            response = s3_client.get_object(Bucket=cloudtrail_bucket, Key=object_key)
+            log_data = gzip.decompress(response["Body"].read()).decode("utf-8")
```
Comment (julianwood) [inline cloudtrail-lambda-dynamo-cdk/src/lib/lambda/populate_dynamo/index.py:28]: When running the test, this line in the function errors:
```
[ERROR] BadGzipFile: Not a gzipped file (b'Te')
Traceback (most recent call last):
  File "/var/task/index.py", line 28, in lambda_handler
    log_data = gzip.decompress(response["Body"].read()).decode("utf-8")
  File "/var/lang/lib/python3.10/gzip.py", line 557, in decompress
    return f.read()
  File "/var/lang/lib/python3.10/gzip.py", line 301, in read
    return self._buffer.read(size)
  File "/var/lang/lib/python3.10/_compression.py", line 118, in readall
    while data := self.read(sys.maxsize):
  File "/var/lang/lib/python3.10/gzip.py", line 488, in read
    if not self._read_gzip_header():
  File "/var/lang/lib/python3.10/gzip.py", line 436, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
```

**Comment by mialarconchong** [comment]
@julianwood Changes have been pushed, thanks for reviewing!

**Pull request merged**